test/common/network/connection_impl_test.cc

```diff
@@ -987,21 +985,32 @@ TEST_P(ConnectionImplTest, WriteWithWatermarks) {
   // connection_impl, and try an immediate drain inside of write() to avoid thrashing here.
   EXPECT_CALL(client_callbacks_, onAboveWriteBufferHighWatermark());
   EXPECT_CALL(client_callbacks_, onBelowWriteBufferLowWatermark());
+
   client_connection_->write(first_buffer_to_write, false);
   dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
   EXPECT_EQ(data_to_write, data_written);
 
-  // Now do the write again, but this time configure buffer_ to reject the write
-  // with errno set to EAGAIN via failWrite(). This should result in going above the high
+  // Now do the write again, but this time configure os_sys_calls to reject the write
+  // with errno set to EAGAIN. This should result in going above the high
   // watermark and not returning.
   Buffer::OwnedImpl second_buffer_to_write(data_to_write);
   EXPECT_CALL(*client_write_buffer_, move(_))
       .WillRepeatedly(DoAll(AddBufferToStringWithoutDraining(&data_written),
                             Invoke(client_write_buffer_, &MockWatermarkBuffer::baseMove)));
-  EXPECT_CALL(*client_write_buffer_, write(_))
-      .WillOnce(Invoke([&](IoHandle& io_handle) -> Api::IoCallUint64Result {
+  NiceMock<Api::MockOsSysCalls> os_sys_calls;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
+  EXPECT_CALL(os_sys_calls, writev(_, _, _))
+      .WillOnce(Invoke([&](os_fd_t, const iovec*, int) -> Api::SysCallSizeResult {
+        dispatcher_->exit();
+        return {-1, SOCKET_ERROR_AGAIN};
+      }));
+  bool wait = true;
+  ON_CALL(client_callbacks_, onEvent(ConnectionEvent::RemoteClose))
+      .WillByDefault(Invoke([&](Network::ConnectionEvent) -> void { wait = false; }));
+  ON_CALL(server_callbacks_, onEvent(ConnectionEvent::RemoteClose))
+      .WillByDefault(Invoke([&](Network::ConnectionEvent) -> void {
+        wait = false;
```

Comment:
This is just scaffolding meant to help figure out why windows tests fail. Somehow it is hit, i.e., client closes connection unexpectedly on the server, although I'm not sure what triggers it because linux tests pass without issue. Could be that the changes in the PR invalidate some windows specific assumptions, but I can't tell which from the ci logs. @davinci26, in case you have some time to take a look at this, any idea what it could be? 

@mattklein123, I guess we don't yet have any access to windows vms to do proper debugging, do we?

Reply:
I have checked out your branch and looking into it.

Reply:
Thanks a lot! Let me know if there's anything I could help with.